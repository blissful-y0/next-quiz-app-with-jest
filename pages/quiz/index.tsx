import Quiz from "../../src/component/quiz/quiz.container";

function QuizPage() {
  return <Quiz />;
}

export default QuizPage;
