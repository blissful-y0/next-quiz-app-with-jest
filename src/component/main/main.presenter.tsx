import {
  PageWrapper,
  Question,
  StartQuizButton,
  TextArea,
  UserNameInput,
} from "./main.style";
import Link from "next/link";
import { ChangeEvent } from "react";

interface IProps {
  onChangeInput: (event: ChangeEvent<HTMLInputElement>) => void;
}

function MainUI({ onChangeInput }: IProps) {
  return (
    <PageWrapper>
      <Question data-testid="WelcomeToQuiz">
        퀴즈의 세계에 오신 것을 환영합니다!
        <TextArea>당신의 이름을 입력해 주세요...</TextArea>
        <UserNameInput onChange={onChangeInput} />
        <Link href="/quiz">
          <StartQuizButton>퀴즈 시작하기</StartQuizButton>
        </Link>
      </Question>
    </PageWrapper>
  );
}

export default MainUI;
