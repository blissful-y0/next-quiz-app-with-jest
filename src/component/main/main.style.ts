import styled from "@emotion/styled";

export const PageWrapper = styled.div`
  width: 100%;
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background: linear-gradient(
    340deg,
    #009779,
    #00b092,
    #00c9ab,
    #19e2c4,
    #32fbdd
  );
`;

export const Question = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 80%;
  max-width: 1000px;
  height: 70%;
  min-height: 500px;
  color: #009779;
  font-size: 30px;
  font-weight: bold;
  border: 5px solid white;
  border-radius: 25px;
  padding: 1%;
  text-align: center;
  background-color: white;
`;

export const TextArea = styled.div`
  width: 100%;
  font-size: 25px;
  font-weight: normal;
  margin-top: 2%;
  margin-bottom: 2%;
`;

export const UserNameInput = styled.input`
  width: 50%;
  font-size: 40px;
  text-align: center;
  border: 5px solid #009779;
  padding: 2%;
`;

export const StartQuizButton = styled.button`
  margin-top: 3%;
  width: 300px;
  height: 80px;
  background-color: #009779;
  outline: none;
  border: 0px;
  color: white;
  font-size: 40px;
  font-weight: bold;
`;
