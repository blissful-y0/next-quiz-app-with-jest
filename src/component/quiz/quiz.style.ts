import styled from "@emotion/styled";
import { css } from "@emotion/react";

interface IButton {
  isTrue: boolean;
}

export const PageWrapper = styled.div`
  width: 100%;
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background: linear-gradient(
    340deg,
    #009779,
    #00b092,
    #00c9ab,
    #19e2c4,
    #32fbdd
  );
`;

export const Question = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 80%;
  max-width: 1000px;
  height: 15%;
  min-height: 250px;
  color: white;
  font-size: 25px;
  font-weight: bold;
  border: 5px solid white;
  border-radius: 25px;
  padding: 1%;
  text-align: center;
`;

export const IncorrectAnswerOne = styled.button`
  background-color: ${(props) => (props.disabled ? "#F5F5F5" : "white")};
  font-size: 25px;
  font-weight: bold;
  border: 2px solid white;
  border-radius: 25px;
  width: 100%;
  max-width: 800px;
  height: 45px;
  min-height: 45px;
  margin-bottom: 2%;
  cursor: pointer;
  color: ${(props: IButton) =>
    props.isTrue === undefined ? "#009779" : "#ff0000"};
`;
export const IncorrectAnswerTwo = styled.button`
  background-color: ${(props) => (props.disabled ? "#F5F5F5" : "white")};
  font-size: 25px;
  font-weight: bold;
  border: 2px solid white;
  border-radius: 25px;
  width: 100%;
  max-width: 800px;
  height: 45px;
  min-height: 45px;
  margin-bottom: 2%;
  cursor: pointer;
  color: ${(props: IButton) =>
    props.isTrue === undefined ? "#009779" : "#ff0000"};
`;
export const IncorrectAnswerThree = styled.button`
  background-color: ${(props) => (props.disabled ? "#F5F5F5" : "white")};
  font-size: 25px;
  font-weight: bold;
  border: 2px solid white;
  border-radius: 25px;
  width: 100%;
  max-width: 800px;
  height: 45px;
  min-height: 45px;
  margin-bottom: 2%;
  cursor: pointer;
  color: ${(props: IButton) =>
    props.isTrue === undefined ? "#009779" : "#ff0000"};
`;

export const CorrectAnswer = styled.button`
  background-color: ${(props) => (props.disabled ? "#F5F5F5" : "white")};
  font-size: 25px;
  font-weight: bold;
  border: 2px solid white;
  border-radius: 25px;
  width: 100%;
  max-width: 800px;
  height: 45px;
  min-height: 45px;
  margin-bottom: 2%;
  cursor: pointer;
  color: ${(props: IButton) =>
    props.isTrue === undefined ? "#009779" : "#00ffaa"};
`;

export const AnswerContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 80%;
  margin-top: 3%;
`;

export const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-around;
  align-items: center;
  width: 80%;
`;

export const RetryButton = styled.button`
  text-align: center;
  background-color: #00c9ab;
  color: black;
  font-size: 25px;
  font-weight: bold;
  border: 2px solid #00c9ab;
  border-radius: 25px;
  width: 49%;
  max-width: 800px;
  height: 45px;
  min-height: 45px;
  margin-bottom: 2%;
  cursor: pointer;
  outline: none;
`;

export const NextButton = styled.button`
  text-align: center;
  background-color: #00c9ab;
  color: black;
  font-size: 25px;
  font-weight: bold;
  border: 2px solid #00c9ab;
  border-radius: 25px;
  width: 49%;
  max-width: 800px;
  height: 45px;
  min-height: 45px;
  margin-bottom: 2%;
  cursor: pointer;
  outline: none;
`;
