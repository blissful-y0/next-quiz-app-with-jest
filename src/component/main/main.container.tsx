import { useContext } from "react";
import { GlobalContext } from "../../../pages/_app";
import MainUI from "./main.presenter";
import { ChangeEvent } from "react";

function Main() {
  const { setUserName } = useContext(GlobalContext);

  const onChangeInput = (event: ChangeEvent<HTMLInputElement>) => {
    setUserName(event.target.value);
  };

  return <MainUI onChangeInput={onChangeInput} />;
}

export default Main;
