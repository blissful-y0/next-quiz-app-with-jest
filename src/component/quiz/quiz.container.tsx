import { useRouter } from "next/dist/client/router";
import { useState, useEffect, useContext } from "react";
import { GlobalContext } from "../../../pages/_app";
import { useTimer } from "../../common/libraries/utils";
import QuizUI from "./quiz.presenter";
import { API_URL } from "./quiz.query";

const Quiz = () => {
  const router = useRouter();
  const { data, setData, userName } = useContext(GlobalContext);
  const [index, setIndex] = useState<number>(0);
  const [isSelectable, setIsSelectable] = useState<boolean>(false);
  const [isButtonVisible, setIsButtonVisible] = useState<"visible" | "hidden">(
    "hidden"
  );
  const [answersArray, setAnswersArray] = useState<boolean[]>([]);
  const [totalSeconds, setTotalSeconds] = useState<number[]>([]);

  const { setIsRunning, seconds, setSeconds } = useTimer();

  useEffect(() => {
    fetch(API_URL)
      .then((result) => result.json())
      .then((data) => {
        setData(data.results);
      });
  }, []);

  const onClickAnswers = (event: boolean) => {
    setIsRunning(false);
    setIsButtonVisible("visible");
    setIsSelectable((props) => !props);
    setAnswersArray((prev) => [...prev, event]);
  };

  const onClickNextButton: React.MouseEventHandler<HTMLButtonElement> = () => {
    if (index === 9) return;
    setIsRunning(true);
    setTotalSeconds((prev) => [...prev, seconds]);
    setIsButtonVisible("hidden");
    setIsSelectable((props) => !props);
    setSeconds(1);
    setIndex(index + 1);
  };

  const onClickFinishButton: React.MouseEventHandler<HTMLButtonElement> = (
    event
  ) => {
    if (totalSeconds.length > 10) return;
    setIsRunning(false);
    setTotalSeconds((prev) => [...prev, seconds]);
    const totalTime = totalSeconds.reduce((acc, curr) => acc + curr, 0);
    localStorage.setItem("Answers", `${JSON.stringify(answersArray)}`);
    localStorage.setItem("TotalTimer", `${totalTime}`);
    router.push({
      pathname: "/result",
      query: {
        showResult: userName,
      },
    });
  };

  const onClickRetryButton: React.MouseEventHandler<HTMLButtonElement> = () => {
    if (index === 9) {
      setAnswersArray([]);
      setIsButtonVisible("hidden");
      setIsSelectable(false);
      setIndex(0);
      setSeconds(1);
      setIsRunning(true);
      localStorage.clear();
    } else {
      let retryArray: boolean[] = [...answersArray];
      retryArray.pop();
      setAnswersArray(retryArray);
      setIsButtonVisible("hidden");
      setIsSelectable(false);
      setSeconds(1);
      setIsRunning(true);
    }
  };

  return (
    <QuizUI
      data={data}
      onClickNextButton={onClickNextButton}
      index={index}
      isButtonVisible={isButtonVisible}
      onClickAnswers={onClickAnswers}
      isSelectable={isSelectable}
      onClickFinishButton={onClickFinishButton}
      onClickRetryButton={onClickRetryButton}
      answersArray={answersArray}
    />
  );
};

export default Quiz;
