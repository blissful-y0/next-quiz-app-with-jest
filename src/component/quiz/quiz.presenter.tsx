import {
  PageWrapper,
  Question,
  AnswerContainer,
  NextButton,
  RetryButton,
  ButtonContainer,
  IncorrectAnswerOne,
  IncorrectAnswerTwo,
  IncorrectAnswerThree,
  CorrectAnswer,
} from "./quiz.style";
import MoonLoader from "react-spinners/ClipLoader";
import { IData } from "./quiz.type";

interface IProps {
  data: IData[];
  onClickNextButton: React.MouseEventHandler<HTMLButtonElement>;
  onClickAnswers: (event: boolean) => void;
  index: number;
  isButtonVisible: "visible" | "hidden";
  isSelectable: boolean;
  onClickFinishButton: React.MouseEventHandler<HTMLButtonElement>;
  onClickRetryButton: React.MouseEventHandler<HTMLButtonElement>;
  answersArray: boolean[];
}

function QuizUI({
  data,
  onClickNextButton,
  index,
  isButtonVisible,
  onClickAnswers,
  isSelectable,
  onClickFinishButton,
  onClickRetryButton,
  answersArray,
}: IProps) {
  return (
    <>
      <PageWrapper>
        {data.length === 1 ? (
          <MoonLoader
            size={300}
            loading={true}
            color={"white"}
            speedMultiplier={1}
          />
        ) : (
          <>
            <Question>{data[index]?.question}</Question>
            <AnswerContainer>
              <IncorrectAnswerOne
                //@ts-ignore
                isTrue={answersArray[index]}
                disabled={isSelectable}
                onClick={() => onClickAnswers(false)}
              >
                {data[index]?.incorrect_answers[0]}
              </IncorrectAnswerOne>
              <IncorrectAnswerTwo
                //@ts-ignore
                isTrue={answersArray[index]}
                disabled={isSelectable}
                onClick={() => onClickAnswers(false)}
              >
                {data[index]?.incorrect_answers[1]}
              </IncorrectAnswerTwo>
              <IncorrectAnswerThree
                //@ts-ignore
                isTrue={answersArray[index]}
                disabled={isSelectable}
                onClick={() => onClickAnswers(false)}
              >
                {data[index]?.incorrect_answers[2]}
              </IncorrectAnswerThree>
              <CorrectAnswer
                //@ts-ignore
                isTrue={answersArray[index]}
                disabled={isSelectable}
                onClick={() => onClickAnswers(true)}
              >
                {data[index]?.correct_answer}
              </CorrectAnswer>
              <ButtonContainer style={{ visibility: isButtonVisible }}>
                <RetryButton onClick={onClickRetryButton}>Retry</RetryButton>
                {index === 9 ? (
                  <NextButton onClick={onClickFinishButton}>Finish</NextButton>
                ) : (
                  <NextButton onClick={onClickNextButton}>Next</NextButton>
                )}
              </ButtonContainer>
            </AnswerContainer>
          </>
        )}
      </PageWrapper>
    </>
  );
}

export default QuizUI;
