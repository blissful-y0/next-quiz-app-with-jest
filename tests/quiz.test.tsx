import { render, screen, cleanup } from "@testing-library/react";
import { Question, StartQuizButton } from "../src/component/main/main.style";
import "@testing-library/jest-dom/extend-expect";
import { createSerializer } from "@emotion/jest";
import renderer from "react-test-renderer";
import QuizPage from "../pages/quiz";
import { API_URL } from "../src/component/quiz/quiz.query";

describe("Should render the app without crashing", () => {
  it("Renders the page without crash", () => {
    render(<QuizPage />);
  });

  it("renders homepage unchaged", () => {
    const tree = renderer.create(<QuizPage />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
