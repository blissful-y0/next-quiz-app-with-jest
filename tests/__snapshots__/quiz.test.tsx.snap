// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the app without crashing renders homepage uncahged 1`] = `
<div
  className="css-fsuynv"
>
  <span
    className="css-42ocmx"
  >
    <span
      className="css-u5qs1v"
    />
    <span
      className="css-mt367r"
    />
  </span>
</div>
`;

exports[`Should render the app without crashing renders homepage unchaged 1`] = `
<div
  className="css-fsuynv"
>
  <span
    className="css-42ocmx"
  >
    <span
      className="css-u5qs1v"
    />
    <span
      className="css-mt367r"
    />
  </span>
</div>
`;
